File: julia/phone.jl
module phone

export add, parse

mutable struct Node{T}
    data::T
    iso::String
    child::Vector{Union{Nothing,Node{T}}}
    Node{T}(data,iso) where {T} =(y=new(); y.data=data; y.iso=iso; y.child=Vector{Union{Nothing,Node{T}}}(nothing,10); y)
end

TREE=Node{Int}(0,"root")

function add(cc::String, iso::String)
   root = TREE 
   i::Int = 0
   for ch in cc
    index = ch - '0' + 1
    node = root.child[index]
    if node === nothing
        root.child[index] = Node{Int}(0,"")
        node = root.child[index]
    end
    i+=1
    if i == length(cc)
        node.data = 1
        node.iso = iso
    else
        root = node
    end
   end 
end

function parse(phone::String)
    len::Int = 0
    ld::Int = 0
    ac::Int = 0
    iso::String=""
    root = TREE 
    i::Int = 0
    for ch in phone
     index = ch - '0' + 1
     node = root.child[index]
     if node === nothing
        break
     end
     i += 1
     if node.data == 1
        if len == 0 len = i
        elseif ld == 0 ld = i
        else ac = i
        end
        iso = node.iso
     end
     root = node
    end

    len,ld,ac,iso
end

end

File: ts/phone.ts
class Node {
    public type: Number
    public iso: string
    public child: Node[]
};

export class PhoneNo {
    private tree: Node
    constructor() {
        this.tree = new Node()
    }

    public add(cc: string, iso: string) {
        let root = this.tree
        for (let i = 0; i < cc.length; ++i) {
            let index = cc.charCodeAt(i) - '0'.charCodeAt(0)
            if (!root.child || ! root.child[index]) {
                if (!root.child) root.child = []
                root.child[index] = new Node()
            }
            let node = root.child[index]
            if (i == cc.length - 1) {
                node.type = 1
                node.iso = iso
            } else {
                root = node
            }
        }
    }

    public parse(phone: string): [number,number,number,string] {
        let ln = 0
        let ld = 0
        let ac = 0
        let iso:string = null
        let root = this.tree
        for (let i = 0; i < phone.length; ++i) {
            let index = phone.charCodeAt(i) - '0'.charCodeAt(0)
            if (!root.child || ! root.child[index]) {
                break
            }
            let node = root.child[index]
            if (node.type == 1) {
                if (ln == 0) ln = i + 1
                else if (ld == 0) ld = i + 1
                else ac = i + 1
                iso = node.iso
            }
            root = node
        }
        return [ln,ld,ac,iso]
    }
};
